BSP/STM32F413/Src/BSP_CAN.c

```diff
@@ -123,23 +127,32 @@ uint8_t BSP_CAN_Write(uint32_t id, uint8_t data[8], uint8_t length) {
  * @return  0 if nothing was received so ignore id and data that was received. Any other value indicates data was received and stored.
  */
 uint8_t BSP_CAN_Read(uint32_t *id, uint8_t *data) {
-    if(RxFlag){
+    //if the head and tail are not the same, or they are the same and it is overflowing
+    if((rxFifoHead != rxFifoTail) || (rxFifoHead == rxFifoTail && rxFifoOverFlow == 1)){
 		for(int i = 0; i < 8; i++){
-			data[i] = RxMessage.Data[i];
+			data[i] = rxFifo[rxFifoTail][i];
 		}
+        rxFifoTail = (rxFifoTail + 1)%CAN_FIFO_SIZE; //increment FIFO tail
         *id = RxMessage.StdId;
-		RxFlag = false;
 		return 1;
+        //if the head is not right below the tail, remove error status
+        if ((rxFifoHead + 1)%CAN_FIFO_SIZE != rxFifoTail) rxFifoOverFlow = 0;
```

Comment:
If you overflow by more than one message, I'm pretty sure that you'll lose all of the previous messages this way. Consider the following example.
```
-----
| a | <-- TAIL
| b |
| c |
| d |
| e | <-- HEAD
-----          overflow = 0
```
Say two more messages come in. Then your buffer will look like this:
```
-----
| f | <-- TAIL
| g | <-- HEAD
| c |
| d |
| e |
-----          overflow = 1
```
Now say that you read one message. Then you get something like this:
```
-----
| f | <-- HEAD, TAIL
| g |
| c |
| d |
| e |
-----          overflow = 0
```
Notice how `c`, `d`, `e`, and `f` will never be read, despite still being valid.

The proper way to handle overflow would probably be not with a flag, but rather by erasing the oldest element, i.e. by increment the tail by one before adding the element if the buffer is full. This way, only the oldest messages are lost, rather than messages that we may still want to attend to.

Reply:
I'm kind of confused by the FIFO you drew. If the head is storing data and moving down, and the tail is storing data and moving down, wouldn't that mean that there isn't any data in g,c,d,e? Also, I set the error bit when the head is right before the tail, not when the tail is right before the head so the overflow bit wouldn't be set in the second graph. If I'm reading it wrong then let me know. Basically what I was trying to do was set apart two events. One when the head == tail and the FIFO is empty or when head == tail and the FIFO is full (that's what I used the error bit for). Also, I am storing/reading data, then moving so I'm not sure if there would be data in F.

Reply:
If you want, I can just erase the oldest element by incrementing TAIL when there is overflow. It's definitely easier than this.